File: backend/src/handlers/wallet_sdk.rs
use axum::{
    extract::{Extension, State},
    http::StatusCode,
    Json,
};
use serde_json::{json, Value};
use sqlx::SqlitePool;

use crate::wallet_sdk_service::{WalletSDKService, SendPaymentRequest};

pub async fn create_wallet(
    State(_pool): State<SqlitePool>,
    Extension(_user_id): Extension<String>,
) -> Result<Json<Value>, StatusCode> {
    let service = WalletSDKService::new();
    let response = service.create_wallet();
    
    Ok(Json(json!({
        "public_key": response.public_key,
        "secret_key": response.secret_key,
        "wallet_id": response.wallet_id
    })))
}

pub async fn get_wallet_balance_sdk(
    State(_pool): State<SqlitePool>,
    Extension(_user_id): Extension<String>,
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    let secret_key = payload["secret_key"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;

    let service = WalletSDKService::new();
    
    match service.get_wallet_balance(secret_key).await {
        Ok(response) => Ok(Json(json!({
            "balances": response.balances,
            "public_key": response.public_key
        }))),
        Err(_) => Err(StatusCode::INTERNAL_SERVER_ERROR),
    }
}

pub async fn send_payment_sdk(
    State(_pool): State<SqlitePool>,
    Extension(_user_id): Extension<String>,
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    let secret_key = payload["secret_key"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;
    
    let destination = payload["destination"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;
    
    let amount = payload["amount"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;

    let asset_code = payload["asset_code"].as_str();

    let request = SendPaymentRequest {
        destination: destination.to_string(),
        amount: amount.to_string(),
        asset_code: asset_code.map(|s| s.to_string()),
        memo: payload["memo"].as_str().map(|s| s.to_string()),
    };

    let service = WalletSDKService::new();
    
    match service.send_payment(secret_key, request).await {
        Ok(result) => Ok(Json(json!({
            "transaction_hash": result.transaction_hash,
            "amount": result.amount,
            "asset_code": result.asset_code,
            "destination": result.destination
        }))),
        Err(_) => Err(StatusCode::INTERNAL_SERVER_ERROR),
    }
}

pub async fn fund_testnet_sdk(
    State(_pool): State<SqlitePool>,
    Extension(_user_id): Extension<String>,
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    let secret_key = payload["secret_key"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;

    let service = WalletSDKService::new();
    
    match service.fund_testnet_account(secret_key).await {
        Ok(success) => Ok(Json(json!({
            "success": success,
            "message": if success { "Account funded successfully" } else { "Failed to fund account" }
        }))),
        Err(_) => Err(StatusCode::INTERNAL_SERVER_ERROR),
    }
}

pub async fn create_trustline_sdk(
    State(_pool): State<SqlitePool>,
    Extension(_user_id): Extension<String>,
    Json(payload): Json<Value>,
) -> Result<Json<Value>, StatusCode> {
    let secret_key = payload["secret_key"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;
    
    let asset_code = payload["asset_code"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;
    
    let issuer = payload["issuer"]
        .as_str()
        .ok_or(StatusCode::BAD_REQUEST)?;

    let limit = payload["limit"].as_str();

    let service = WalletSDKService::new();
    
    match service.create_trustline(secret_key, asset_code, issuer, limit).await {
        Ok(tx_hash) => Ok(Json(json!({
            "transaction_hash": tx_hash,
            "asset_code": asset_code,
            "issuer": issuer
        }))),
        Err(_) => Err(StatusCode::INTERNAL_SERVER_ERROR),
    }
}

File: frontend/src/services/walletSDK.ts
const API_BASE_URL = 'https://9c8ea870332a.ngrok-free.app';

interface WalletCreateResponse {
  public_key: string;
  secret_key: string;
  wallet_id: string;
}

interface Balance {
  asset_code: string;
  balance: string;
  limit?: string;
}

interface WalletBalanceResponse {
  balances: Balance[];
  public_key: string;
}

interface PaymentResult {
  transaction_hash: string;
  amount: string;
  asset_code: string;
  destination: string;
}

interface SendPaymentRequest {
  destination: string;
  amount: string;
  asset_code?: string;
  memo?: string;
}

class WalletSDK {
  private getAuthHeaders() {
    const token = localStorage.getItem('novapay_token');
    return token ? { 'Authorization': `Bearer ${token}` } : {};
  }

  async createWallet(): Promise<WalletCreateResponse> {
    const response = await fetch(`${API_BASE_URL}/sdk/wallet/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders()
      }
    });
    
    if (!response.ok) throw new Error('Failed to create wallet');
    return response.json();
  }

  async getWalletBalance(secretKey: string): Promise<WalletBalanceResponse> {
    const response = await fetch(`${API_BASE_URL}/sdk/wallet/balance`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders()
      },
      body: JSON.stringify({ secret_key: secretKey })
    });
    
    if (!response.ok) throw new Error('Failed to get wallet balance');
    return response.json();
  }

  async sendPayment(secretKey: string, payment: SendPaymentRequest): Promise<PaymentResult> {
    const response = await fetch(`${API_BASE_URL}/sdk/wallet/send`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders()
      },
      body: JSON.stringify({
        secret_key: secretKey,
        ...payment
      })
    });
    
    if (!response.ok) throw new Error('Failed to send payment');
    return response.json();
  }

  async fundTestnetAccount(secretKey: string): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${API_BASE_URL}/sdk/wallet/fund-testnet`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders()
      },
      body: JSON.stringify({ secret_key: secretKey })
    });
    
    if (!response.ok) throw new Error('Failed to fund testnet account');
    return response.json();
  }

  async createTrustline(
    secretKey: string,
    assetCode: string,
    issuer: string,
    limit?: string
  ): Promise<{ transaction_hash: string; asset_code: string; issuer: string }> {
    const response = await fetch(`${API_BASE_URL}/sdk/wallet/trustline`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...this.getAuthHeaders()
      },
      body: JSON.stringify({
        secret_key: secretKey,
        asset_code: assetCode,
        issuer,
        limit
      })
    });
    
    if (!response.ok) throw new Error('Failed to create trustline');
    return response.json();
  }

  validateStellarAddress(address: string): boolean {
    return address.length === 56 && address.startsWith('G');
  }

  validateStellarSecret(secret: string): boolean {
    return secret.length === 56 && secret.startsWith('S');
  }

  formatAmount(amount: string | number): string {
    return parseFloat(amount.toString()).toFixed(7);
  }

  stroopsToXLM(stroops: string | number): string {
    return (parseFloat(stroops.toString()) / 10000000).toFixed(7);
  }

  xlmToStroops(xlm: string | number): string {
    return (parseFloat(xlm.toString()) * 10000000).toString();
  }
}

export const walletSDK = new WalletSDK();
export type { WalletCreateResponse, Balance, WalletBalanceResponse, PaymentResult, SendPaymentRequest };